File: src-tauri/src/tray.rs
//! System-tray (menu bar) icon and its menu, plus the app's state holder that
//! keeps the tray alive for the lifetime of the app.

use tauri::menu::{Menu, MenuItem, PredefinedMenuItem};
use tauri::tray::{MouseButton, MouseButtonState, TrayIcon, TrayIconBuilder, TrayIconEvent};
use tauri::{App, AppHandle, Manager};

/// Managed state that owns the tray icon so it is not dropped (and thus hidden).
pub struct TrayState {
    _tray: TrayIcon,
}

/// Show and focus the main window (created lazily by the boot flow).
fn show_main(app: &AppHandle) {
    if let Some(win) = app.get_webview_window("main") {
        let _ = win.show();
        let _ = win.set_focus();
    }
}

/// Reload the dsh web UI in the main window without restarting dsh.
fn reload_main(app: &AppHandle) {
    if let Some(win) = app.get_webview_window("main") {
        let _ = win.eval("window.location.reload()");
    }
}

/// Build the tray icon + menu and register it in the app's managed state.
pub fn setup(app: &mut App) -> Result<(), Box<dyn std::error::Error>> {
    let show = MenuItem::with_id(app, "show", "Show dsh Window", true, None::<&str>)?;
    let reload = MenuItem::with_id(app, "reload", "Reload Page", true, None::<&str>)?;
    let restart = MenuItem::with_id(app, "restart", "Restart", true, None::<&str>)?;
    let quit = MenuItem::with_id(app, "quit", "Quit dsh Desktop", true, None::<&str>)?;
    let sep = PredefinedMenuItem::separator(app)?;

    let menu = Menu::with_items(app, &[&show, &reload, &restart, &sep, &quit])?;

    let icon = app.default_window_icon().map(|i| i.clone());

    let mut builder = TrayIconBuilder::with_id("dsh-tray")
        .menu(&menu)
        .show_menu_on_left_click(false)
        .on_menu_event(|app_handle, event| match event.id().as_ref() {
            "show" => show_main(app_handle),
            "reload" => reload_main(app_handle),
            "restart" => app_handle.request_restart(),
            "quit" => app_handle.exit(0),
            _ => {}
        })
        .on_tray_icon_event(|tray, event| {
            if let TrayIconEvent::Click {
                button: MouseButton::Left,
                button_state: MouseButtonState::Up,
                ..
            } = event
            {
                show_main(tray.app_handle());
            }
        });

    if let Some(icon) = icon {
        builder = builder.icon(icon);
    }

    let tray = builder.build(app)?;
    app.manage(TrayState { _tray: tray });
    Ok(())
}

File: src/tray.js
'use strict'

// System-tray (menu bar) icon and its menu (Electron port of tray.rs).

const { Tray, Menu, nativeImage } = require('electron')
const fs = require('node:fs')

/** Show and focus the main window (created lazily by the boot flow). */
function showMain(app, getWindow) {
  const win = getWindow()
  if (win && !win.isDestroyed()) {
    win.show()
    win.focus()
  }
}

/** Reload the dsh web UI without restarting dsh. */
function reloadMain(getWindow) {
  const win = getWindow()
  if (win && !win.isDestroyed()) win.webContents.reload()
}

/**
 * Build the tray icon + menu.
 * @param {import('electron').App} app
 * @param {{ getWindow: () => Electron.BrowserWindow | undefined }} ctx
 * @param {string} trayIconPath - path to the 16x16 tray PNG (with `@2x` sibling).
 */
function setup(app, ctx, trayIconPath) {
  const template = [
    { label: 'Show dsh Window', click: () => showMain(app, ctx.getWindow) },
    { label: 'Reload Page', click: () => reloadMain(ctx.getWindow) },
    { label: 'Restart', click: () => { app.relaunch(); app.quit() } },
    { type: 'separator' },
    { label: 'Quit dsh Desktop', click: () => app.quit() },
  ]
  const menu = Menu.buildFromTemplate(template)

  let image
  if (trayIconPath && fs.existsSync(trayIconPath)) {
    // Show the app icon as-is (colored), matching the Tauri tray behavior.
    image = nativeImage.createFromPath(trayIconPath)
  }

  const tray = new Tray(image)
  tray.setToolTip('dsh Desktop')
  tray.setContextMenu(menu)
  // Left-click brings the window back.
  tray.on('click', () => showMain(app, ctx.getWindow))
  return tray
}

module.exports = { setup, showMain }
